File: clients/dotnet/Connecto.Client/ConnectoClient.cs
using System.Net.Http.Json;
using Microsoft.AspNetCore.SignalR.Client;

namespace Connecto.Client;

public class ConnectoClient : IAsyncDisposable
{
    private readonly string _baseUrl;
    private readonly HttpClient _http;
    private string? _token;
    private HubConnection? _dataHub;
    private HubConnection? _collectionHub;

    public bool IsConnected =>
        _dataHub?.State == HubConnectionState.Connected &&
        _collectionHub?.State == HubConnectionState.Connected;

    public ConnectoClient(string url = "http://localhost:5043")
    {
        _baseUrl = url.TrimEnd('/');
        _http = new HttpClient { BaseAddress = new Uri(_baseUrl) };
    }

    // ── Auth ────────────────────────────────────────────────────

    public async Task<AuthRes> Login(LoginReq req)
    {
        var res = await _http.PostAsJsonAsync("/api/auth/login", req);
        await ThrowIfNotOk(res);
        var auth = (await res.Content.ReadFromJsonAsync<AuthRes>())!;
        _token = auth.Token;
        return auth;
    }

    public async Task<AuthRes> Register(RegisterReq req)
    {
        var res = await _http.PostAsJsonAsync("/api/auth/register", req);
        await ThrowIfNotOk(res);
        var auth = (await res.Content.ReadFromJsonAsync<AuthRes>())!;
        _token = auth.Token;
        return auth;
    }

    // ── Connection ──────────────────────────────────────────────

    public async Task Connect()
    {
        if (_token is null) throw new InvalidOperationException("Not authenticated.");
        _dataHub = BuildHub("/data-stream");
        _collectionHub = BuildHub("/collection-stream");

        _dataHub.On<string>("HubError", err => Console.Error.WriteLine($"[connecto] {err}"));
        _collectionHub.On<string>("HubError", err => Console.Error.WriteLine($"[connecto] {err}"));

        await Task.WhenAll(_dataHub.StartAsync(), _collectionHub.StartAsync());
    }

    // ── Collections ─────────────────────────────────────────────

    public Task<List<string>> ListTables() =>
        RequireCollection().InvokeAsync<List<string>>("ListTables");

    public Task CreateTable(string name)
    {
        var tcs = new TaskCompletionSource<string>();
        RequireCollection().On<string>("TableCreated", table => tcs.TrySetResult(table));
        RequireCollection().InvokeAsync("CreateTable", name);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    public Task DeleteTable(string name)
    {
        var tcs = new TaskCompletionSource<string>();
        RequireCollection().On<string>("TableDeleted", table => tcs.TrySetResult(table));
        RequireCollection().InvokeAsync("DeleteTable", name);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    // ── Data ────────────────────────────────────────────────────

    public Task Subscribe(string tableName) =>
        RequireData().InvokeAsync("SubscribeToTable", tableName);

    public Task Unsubscribe(string tableName) =>
        RequireData().InvokeAsync("UnsubscribeFromTable", tableName);

    public Task<List<FlexMap>> GetAllRecords(string tableName)
    {
        var tcs = new TaskCompletionSource<List<FlexMap>>();
        RequireData().On<string, List<FlexMap>>("EntitiesRequested", (_, records) =>
            tcs.TrySetResult(records));
        RequireData().InvokeAsync("GetAllRecords", tableName);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    public Task<FlexMap?> GetRecord(string tableName, Guid id)
    {
        var tcs = new TaskCompletionSource<FlexMap?>();
        RequireData().On<string, FlexMap?>("EntityRequested", (_, record) =>
            tcs.TrySetResult(record));
        RequireData().InvokeAsync("GetRecord", tableName, id);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    public Task Upsert(string tableName, FlexMap data)
    {
        var tcs = new TaskCompletionSource<FlexMap>();
        if (data.ContainsKey("id"))
        {
            RequireData().On<string, FlexMap>("EntityUpdated", (_, record) =>
                tcs.TrySetResult(record));
        }
        else
        {
            RequireData().On<string, FlexMap>("EntityCreated", (_, record) =>
                tcs.TrySetResult(record));
        }
        RequireData().InvokeAsync("UpsertDataRecord", tableName, data);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    public Task DeleteRecord(string tableName, Guid id)
    {
        var tcs = new TaskCompletionSource<string>();
        RequireData().On<string, string>("EntityDeleted", (_, deletedId) =>
            tcs.TrySetResult(deletedId));
        RequireData().InvokeAsync("DeleteRecord", tableName, id);
        return tcs.Task.WaitAsync(TimeSpan.FromSeconds(30));
    }

    // ── Realtime Listeners ──────────────────────────────────────

    public void OnEntityCreated(Action<string, FlexMap> handler) =>
        RequireData().On("EntityCreated", handler);

    public void OnEntityUpdated(Action<string, FlexMap> handler) =>
        RequireData().On("EntityUpdated", handler);

    public void OnEntityDeleted(Action<string, string> handler) =>
        RequireData().On("EntityDeleted", handler);

    public void OnTableCreated(Action<string> handler) =>
        RequireCollection().On("TableCreated", handler);

    public void OnTableDeleted(Action<string> handler) =>
        RequireCollection().On("TableDeleted", handler);

    public void OnError(Action<string> handler) =>
        RequireData().On("HubError", handler);

    // ── IDisposable ─────────────────────────────────────────────

    public async ValueTask DisposeAsync()
    {
        if (_dataHub is not null) await _dataHub.DisposeAsync();
        if (_collectionHub is not null) await _collectionHub.DisposeAsync();
        _http.Dispose();
    }

    // ── Internals ───────────────────────────────────────────────

    private HubConnection RequireData() =>
        _dataHub ?? throw new InvalidOperationException("Not connected.");

    private HubConnection RequireCollection() =>
        _collectionHub ?? throw new InvalidOperationException("Not connected.");

    private HubConnection BuildHub(string path) =>
        new HubConnectionBuilder()
            .WithUrl($"{_baseUrl}{path}", opts =>
            {
                opts.AccessTokenProvider = () => Task.FromResult(_token)!;
            })
            .WithAutomaticReconnect()
            .Build();

    private static async Task ThrowIfNotOk(HttpResponseMessage res)
    {
        if (!res.IsSuccessStatusCode)
        {
            var err = await res.Content.ReadFromJsonAsync<HubError>();
            throw new HttpRequestException(err?.Message ?? res.ReasonPhrase);
        }
    }
}

File: clients/typescript/src/connecto-client.ts
import {
  HubConnectionBuilder,
  HubConnection,
  HttpTransportType,
  LogLevel,
} from "@microsoft/signalr";
import type { AuthRes, LoginReq, RegisterReq, FlexMap, UserRes, HubError } from "./types";

export type { AuthRes, LoginReq, RegisterReq, FlexMap, UserRes, HubError };

const EVENTS = {
  EntityCreated: "EntityCreated",
  EntityUpdated: "EntityUpdated",
  EntityDeleted: "EntityDeleted",
  EntityRequested: "EntityRequested",
  EntitiesRequested: "EntitiesRequested",
  TablesRequested: "TablesRequested",
  TableCreated: "TableCreated",
  TableDeleted: "TableDeleted",
  HubError: "HubError",
} as const;

export class ConnectoClient {
  private readonly baseUrl: string;
  private token: string | null = null;
  private dataHub: HubConnection | null = null;
  private collectionHub: HubConnection | null = null;

  constructor(url: string = "http://localhost:5043") {
    this.baseUrl = url.replace(/\/$/, "");
  }

  get isConnected(): boolean {
    return (
      this.dataHub?.state === "Connected" &&
      this.collectionHub?.state === "Connected"
    );
  }

  // ── Auth (REST) ──────────────────────────────────────────────

  async login(req: LoginReq): Promise<AuthRes> {
    const res = await this.post("/api/auth/login", req);
    this.token = res.token;
    return res;
  }

  async register(req: RegisterReq): Promise<AuthRes> {
    const res = await this.post("/api/auth/register", req);
    this.token = res.token;
    return res;
  }

  // ── Connection ───────────────────────────────────────────────

  async connect(): Promise<void> {
    if (!this.token) throw new Error("Not authenticated. Call login() or register() first.");
    await Promise.all([this.connectDataHub(), this.connectCollectionHub()]);
  }

  async disconnect(): Promise<void> {
    await Promise.all([
      this.dataHub?.stop(),
      this.collectionHub?.stop(),
    ]);
    this.dataHub = null;
    this.collectionHub = null;
  }

  // ── Collection Operations ────────────────────────────────────

  async listTables(): Promise<string[]> {
    const hub = this.requireCollectionHub();
    return new Promise((resolve, reject) => {
      const handler = (tables: string[]) => {
        hub.off(EVENTS.TablesRequested, handler);
        resolve(tables);
      };
      hub.on(EVENTS.TablesRequested, handler);
      hub.invoke("ListTables").catch(reject);
    });
  }

  async createTable(name: string): Promise<string> {
    const hub = this.requireCollectionHub();
    return new Promise((resolve, reject) => {
      const onCreated = (tableName: string) => {
        hub.off(EVENTS.TableCreated, onCreated);
        resolve(tableName);
      };
      hub.on(EVENTS.TableCreated, onCreated);
      hub.invoke("CreateTable", name).catch(reject);
    });
  }

  async deleteTable(name: string): Promise<string> {
    const hub = this.requireCollectionHub();
    return new Promise((resolve, reject) => {
      const onDeleted = (tableName: string) => {
        hub.off(EVENTS.TableDeleted, onDeleted);
        resolve(tableName);
      };
      hub.on(EVENTS.TableDeleted, onDeleted);
      hub.invoke("DeleteTable", name).catch(reject);
    });
  }

  // ── Data Operations ──────────────────────────────────────────

  async subscribe(tableName: string): Promise<void> {
    await this.requireDataHub().invoke("SubscribeToTable", tableName);
  }

  async unsubscribe(tableName: string): Promise<void> {
    await this.requireDataHub().invoke("UnsubscribeFromTable", tableName);
  }

  async getAllRecords(tableName: string): Promise<FlexMap[]> {
    const hub = this.requireDataHub();
    return new Promise((resolve, reject) => {
      const handler = (_table: string, records: FlexMap[]) => {
        hub.off(EVENTS.EntitiesRequested, handler);
        resolve(records);
      };
      hub.on(EVENTS.EntitiesRequested, handler);
      hub.invoke("GetAllRecords", tableName).catch(reject);
    });
  }

  async getRecord(tableName: string, id: string): Promise<FlexMap> {
    const hub = this.requireDataHub();
    return new Promise((resolve, reject) => {
      const handler = (_table: string, record: FlexMap) => {
        hub.off(EVENTS.EntityRequested, handler);
        resolve(record);
      };
      hub.on(EVENTS.EntityRequested, handler);
      hub.invoke("GetRecord", tableName, id).catch(reject);
    });
  }

  async upsert(tableName: string, data: FlexMap): Promise<FlexMap> {
    const hub = this.requireDataHub();
    if (data.id) {
      return new Promise((resolve, reject) => {
        const handler = (_table: string, record: FlexMap) => {
          hub.off(EVENTS.EntityUpdated, handler);
          resolve(record);
        };
        hub.on(EVENTS.EntityUpdated, handler);
        hub.invoke("UpsertDataRecord", tableName, data).catch(reject);
      });
    }

    return new Promise((resolve, reject) => {
      const handler = (_table: string, record: FlexMap) => {
        hub.off(EVENTS.EntityCreated, handler);
        resolve(record);
      };
      hub.on(EVENTS.EntityCreated, handler);
      hub.invoke("UpsertDataRecord", tableName, data).catch(reject);
    });
  }

  async deleteRecord(tableName: string, id: string): Promise<string> {
    const hub = this.requireDataHub();
    return new Promise((resolve, reject) => {
      const handler = (_table: string, deletedId: string) => {
        hub.off(EVENTS.EntityDeleted, handler);
        resolve(deletedId);
      };
      hub.on(EVENTS.EntityDeleted, handler);
      hub.invoke("DeleteRecord", tableName, id).catch(reject);
    });
  }

  // ── Realtime Listeners ───────────────────────────────────────

  onEntityCreated(cb: (table: string, record: FlexMap) => void): void {
    this.requireDataHub().on(EVENTS.EntityCreated, cb);
  }

  onEntityUpdated(cb: (table: string, record: FlexMap) => void): void {
    this.requireDataHub().on(EVENTS.EntityUpdated, cb);
  }

  onEntityDeleted(cb: (table: string, id: string) => void): void {
    this.requireDataHub().on(EVENTS.EntityDeleted, cb);
  }

  onTableCreated(cb: (table: string) => void): void {
    this.requireCollectionHub().on(EVENTS.TableCreated, cb);
  }

  onTableDeleted(cb: (table: string) => void): void {
    this.requireCollectionHub().on(EVENTS.TableDeleted, cb);
  }

  onError(cb: (error: HubError) => void): void {
    this.requireDataHub().on(EVENTS.HubError, cb);
    this.requireCollectionHub().on(EVENTS.HubError, cb);
  }

  // ── Internals ────────────────────────────────────────────────

  private requireDataHub(): HubConnection {
    if (!this.dataHub) throw new Error("Not connected. Call connect() first.");
    return this.dataHub;
  }

  private requireCollectionHub(): HubConnection {
    if (!this.collectionHub) throw new Error("Not connected. Call connect() first.");
    return this.collectionHub;
  }

  private async connectDataHub(): Promise<void> {
    this.dataHub = this.buildHub("/data-stream");
    this.setupHubErrorHandler(this.dataHub);
    await this.dataHub.start();
  }

  private async connectCollectionHub(): Promise<void> {
    this.collectionHub = this.buildHub("/collection-stream");
    this.setupHubErrorHandler(this.collectionHub);
    await this.collectionHub.start();
  }

  private buildHub(path: string): HubConnection {
    return new HubConnectionBuilder()
      .withUrl(`${this.baseUrl}${path}`, {
        accessTokenFactory: () => this.token!,
        transport: HttpTransportType.WebSockets,
      })
      .configureLogging(LogLevel.Warning)
      .withAutomaticReconnect()
      .build();
  }

  private setupHubErrorHandler(hub: HubConnection): void {
    hub.on(EVENTS.HubError, (err: HubError) => {
      console.error("[connecto]", err.message);
    });
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({ message: res.statusText }));
      throw new Error((err as { message?: string }).message ?? res.statusText);
    }
    return res.json();
  }
}
